lib/authlogic/session/base.rb

```diff
@@ -1849,6 +1881,34 @@ def params_key
         build_key(self.class.params_key)
       end
 
+      def headers_credentials
+        controller.headers[headers_key]
+      end
+
+      def headers_enabled?
+        if !headers_credentials || !klass.column_names.include?("single_access_token")
+          return false
+        end
+        if controller.responds_to_single_access_allowed?
+          return controller.single_access_allowed?
+        end
+        headers_enabled_by_allowed_request_types?
+      end
```

Comment:
Please reduce duplication here by extracting the common bits into a new method like `single_access_token_enabled?`.

```
def headers_enabled?
  headers_credentials && single_access_token_enabled?
end

def params_enabled?
  params_credentials && single_access_token_enabled?
end
```